DabTester.py

```diff
@@ -1,54 +1,68 @@
-        self.dab_client.request(device_id,operation,request)
+    def execute_cmd(self,device_id,dab_dab_request_body_topic,dab_request_body="{}"):
+        self.dab_client.request(device_id,dab_dab_request_body_topic,dab_request_body)
         if self.dab_client.last_error_code() == 200:
             return 0
         else:
             return 1
     
-    def Test_Case(self,device_id,test_case):
-        (operation,request,extra_function,params)=test_case
-        print("\ntesting", operation, " ", request, "... ", end='', flush=True)
+    def Test_Case(self, device_id, test_case):
+        (dab_dab_request_body_topic, dab_request_body, validate_output_function, expected_response_code)=test_case
+        test_result = TestResult(device_id, dab_dab_request_body_topic, dab_request_body, "UNKNOWN", "")
+        print("\ntesting", dab_dab_request_body_topic, " ", dab_request_body, "... ", end='', flush=True)
         start = datetime.datetime.now()
-        if self.execute_cmd(device_id,operation,request) == 0:
+        if self.execute_cmd(device_id, dab_dab_request_body_topic, dab_request_body) == 0:
             end = datetime.datetime.now()
             duration = end - start
             durationInMs = int(duration.total_seconds() * 1000)
-            if extra_function(durationInMs,params) == True:
+            if validate_output_function(durationInMs, expected_response_code) == True:
                 print("\033[1;32m[ PASS ]\033[0m")
+                test_result.test_result = "PASS"
             else:
                 print("\033[1;31m[ FAILED ]\033[0m")
+                test_result.test_result = "FAILED"
         else:
             print('\033[1;31m[ ',end='')
             print("Error",self.dab_client.last_error_code(),': ',end='')
             self.dab_client.last_error_msg()
             print(' ]\033[0m')
         if ((self.verbose == True)):
             print(self.dab_client.response())
+        test_result.response = self.dab_client.response()
+        return test_result
 
-    def Test_All(self,device_id,Test_Set):
-        for operation in Test_Set:
-            self.Test_Case(device_id,operation)
+    def Test_All(self, suite_name, device_id, Test_Set, test_result_output_path):
+        result_list = TestSuite([], suite_name)
+        for dab_dab_request_body_topic in Test_Set:
+            result_list.test_result_list.append(self.Test_Case(device_id, dab_dab_request_body_topic))
             sleep(5)
+        if (len(test_result_output_path) == 0):
+            test_result_output_path = f"./test_result/{suite_name}.json"
+        file_dump = jsons.dumps(result_list, indent = 4)
+        with open(test_result_output_path, "w") as outfile:
+                outfile.write(file_dump)
+
 
     def Close(self):
         self.dab_client.disconnect()
         
 def Default_Test(durationInMs=0, expectedLatencyMs=0):
     sleep(0.2)
-    print("\nOperation Latency, Expected: ", expectedLatencyMs, " ms, Actual: ", durationInMs, " ms\n")
```

Comment:
Can we bring this context into the test result as well? Currently we know what failed but we dont know why e.g. if the failure reason is latency, we wont know it unless we see test logs. Should we either bring these logs as part of the test_result or add these to a supplementary test log file?